apps/guardian/src-tauri/src/main.rs

```diff
@@ -1,130 +1,426 @@
+        }
+    }
+}
+
+/// Holds the currently selected account and corresponding password
+struct ActiveGuardian {
+    /// The guardian data for the active account
+    guardian: GuardianData,
     /// The password used for encryption with the selected account
-    #[allow(dead_code)] // TODO: remove when it is used
     password: Password,
 }
 
+/// The type of managed state for the active guardian
 #[derive(Default)]
-struct ActiveAccountState(Mutex<Option<ActiveAccount>>);
+struct ActiveGuardianState(Mutex<Option<ActiveGuardian>>);
 
-/// Describes possible errors when importing an account.
-#[derive(thiserror::Error, Debug)]
-enum ImportWalletError {
-    /// An error happened while trying to write to disk
-    #[error("Could not save account: {0}")]
-    Write(#[from] std::io::Error),
-    /// Account has already been imported
-    #[error("Account already found in application")]
-    Duplicate,
+/// The collective state kept for all guardians
+#[derive(Default)]
+struct GuardiansState(Mutex<contract::GuardiansState>);
+
+/// The necessary election guard configuration to construct election guard
+/// entities.
+#[derive(Clone)]
+struct ElectionGuardConfig {
+    /// The election manifest
+    #[allow(dead_code)] // TODO: remove when it is used
+    manifest: ElectionManifest,
+    /// The election parameters
+    parameters: ElectionParameters,
 }
 
-impl serde::Serialize for ImportWalletError {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: serde::Serializer, {
-        serializer.serialize_str(self.to_string().as_ref())
+struct ElectionContractMarker;
+type ElectionClient = ContractClient<ElectionContractMarker>;
+
+#[derive(Clone)]
+struct ConnectionConfig {
+    node:     v2::Client,
+    contract: ElectionClient,
+}
+
+async fn get_resource_checked<J: DeserializeOwned>(url: &ChecksumUrl) -> Result<J, Error> {
+    let data = reqwest::get(url.url.clone()).await?.bytes().await?;
+
+    let hash = HashSha2256(sha2::Sha256::digest(&data).into());
+    if url.hash != hash {
+        return Err(anyhow!(
+            "Verification of remote resource at {} failed. Expected checksum {} did not match \
+             computed hash {}.",
+            url.url,
+            url.hash,
+            hash
+        )
+        .into());
     }
+
+    // It's fair to assume that data integrity check means that it can also be
+    // deserialized
+    let result = serde_json::from_slice(&data).unwrap();
+    Ok(result)
 }
 
+impl ConnectionConfig {
+    /// Creates a connection to a concordium node and a contract client. This
+    /// function panics if the necessary environment variables are not set.
+    async fn try_create_from_env() -> Result<Self, Error> {
+        #[allow(clippy::option_env_unwrap)] // To avoid cargo check errors in editor
+        let endpoint_var = option_env!("CCD_ELECTION_NODE")
+            .expect(r#"Expected environment variable "CCD_ELECTION_NODE" to be defined"#); // We
```

Comment:
The rationale is that `option_env!` bakes the environment variable into the application at build time, so this will never happen at runtime unless improper build time configuration is supplied.

Is it not OK to panic in these circumstances? I would think it makes more sense to reserve `Result::Error` for stuff happening at runtime due to dynamic conditions.

Maybe the best option would be to use `env!` after all, and live with the editor errors (or set the environment variables) to get the error at build time? What do you think?

Reply:
I do agree that doing this at compile time is the right thing to do, it's just that it's very odd to use option_env and then immediately unwrap. That's what env is for in my view and it's nice to get a compile time error when you've forgotten to set these variables.

If these compile errors are too annoying I'd recommend adding a local .cargo/config.toml file
```
[env]
CCD_ELECTION_NODE = "https://grpc.testnet.concordium.com:20000"
```

Reply:
Didn't know you could do this! Thanks 😄